modules/memfault-firmware-sdk/memfault_gnss_metrics.c

```diff
@@ -0,0 +1,141 @@
+//              callback and register a wrapper callback.
+//        - location_request() -- initiates a request. Intercept it's result, and
+//              start a session if the request is successful.
+int __real_location_init(location_event_handler_t event_handler);
+int __wrap_location_init(location_event_handler_t event_handler)
+{
+	real_location_event_handler = event_handler;
+	return __real_location_init(memfault_location_event_handler);
+}
+
+// Ensure the substituted function signature matches the original function
+_Static_assert(__builtin_types_compatible_p(__typeof__(&location_init),
+					    __typeof__(&__wrap_location_init)) &&
+		       __builtin_types_compatible_p(__typeof__(&location_init),
+						    __typeof__(&__real_location_init)),
+	       "Error: Wrapped function does not match original location_init function signature");
+
+// Also wrap the location request to know when to start a session
+int __real_location_request(const struct location_config *config);
+int __wrap_location_request(const struct location_config *config)
+{
+	int result = __real_location_request(config);
+
+	if (result == 0) {
+		memfault_gnss_metrics_start_fix_session();
+	}
+	return result;
+}
+
+// Ensure the substituted function signature matches the original function
+_Static_assert(
+	__builtin_types_compatible_p(__typeof__(&location_request),
+				     __typeof__(&__wrap_location_request)) &&
+		__builtin_types_compatible_p(__typeof__(&location_request),
+					     __typeof__(&__real_location_request)),
+	"Error: Wrapped function does not match original location_request function signature");
+
+static void memfault_location_event_handler(const struct location_event_data *event_data)
+{
+	if (event_data->id == LOCATION_EVT_LOCATION || event_data->id == LOCATION_EVT_TIMEOUT ||
+	    event_data->id == LOCATION_EVT_ERROR || event_data->id == LOCATION_EVT_RESULT_UNKNOWN) {
+		memfault_gnss_metrics_stop_fix_session(event_data);
+	}
+	real_location_event_handler(event_data);
+}
+
+/**
+ * @brief Start a GNSS fix session.
+ *
+ */
+static void memfault_gnss_metrics_start_fix_session(void)
+{
+	if (session_in_progress) {
+		return;
+	}
+
+	session_in_progress = true;
+	LOG_DBG("Starting GNSS session");
+	MEMFAULT_METRICS_SESSION_START(ncs_gnss);
+	MEMFAULT_METRIC_ADD(ncs_gnss_fix_request_count, 1);
+}
+
+/**
+ * @brief Stop a GNSS fix session.
+ *
+ * @param event_data pointer to location event data
+ */
+static void memfault_gnss_metrics_stop_fix_session(const struct location_event_data *event_data)
+{
+	if (!session_in_progress) {
+		return;
+	}
+
+	uint32_t session_time_ms;
+	memfault_metrics_heartbeat_timer_read(
+		MEMFAULT_METRICS_KEY(MEMFAULT_METRICS_SESSION_TIMER_NAME(ncs_gnss)),
+		&session_time_ms);
+
+	switch (event_data->id) {
+	case LOCATION_EVT_LOCATION:
+		LOG_DBG("Stopping GNSS session, fix data acquired");
+
+		MEMFAULT_METRIC_SET_UNSIGNED(ncs_gnss_time_to_fix_ms, session_time_ms);
```

Comment:
Is session time the only thing missing from the location event information that you would like to store? I think we could provide the time in data details. And we need to discuss what time do you actually want, i.e., is it only the time GNSS is active that we are interested in here.

Reply:
I think the overall fix time is useful, so if GNSS fails, how long does it take to eventually get a fix if it occurs before the timeout. I can measure that through the session, but it would be convenient if provided via the location data as a new parameter, say `search_time_ms`.

Since you're asking, for GNSS-specific metrics, I would also be interested in the time to the first fix + the associated accuracy and number of satellites. This will of course be the same as the overall metrics recorded if the location accuracy is configured to be normal, but different if the location accuracy is high. This could give insight into the performance on the first fix vs. the final fix. An option here is to add a second `location_data_details_gnss` struct to the `location_data_details` struct called `gnss_first_fix`, and then add a `fix_time_ms` member to the `location_data_details_gnss` struct.

What do you think?

Reply:
> Since you're asking, for GNSS-specific metrics, I would also be interested in the time to the first fix + the associated accuracy and number of satellites. This will of course be the same as the overall metrics recorded if the location accuracy is configured to be normal, but different if the location accuracy is high. This could give insight into the performance on the first fix vs. the final fix.

Any metrics related to the high accuracy mode might be too detailed at this point. I'm not expecting majority of the applications to use that mode but rather the normal accuracy which is the default.
Also, if we collect information on the first fix and final fix, we cannot really use the information for anything. If the plan is to see how the accuracy changes, I'm not sure what conclusions you would make out of that because we don't know the real accuracy compared to the real location of the device. We just have this estimated accuracy.

Reply:
In the event handler, you should check the location_event_data.method to see which method was used and this way it would be easy to track GNSS, cellular and WiFi positioning information separately. And if we would only want track GNSS, you should check that the method is GNSS and ignore others.

Failure reporting is more complex though as a successful location event might still mean that there was a failing GNSS positioning and a successful cellular positioning.

Reply:
> If the plan is to see how the accuracy changes, I'm not sure what conclusions you would make out of that

Yes, the plan would be to see how accuracy changes. If they configure it in high accuracy mode with x number of fixes to get before returning the location, they could see if there is any meaningful accuracy difference between the first and last fix, and if not, reduce the number of fixes. Keeping the GNSS active is a high-power operation, so extra time should be worth the cost.

> we don't know the real accuracy compared to the real location of the device. We just have this estimated accuracy.

Right, I completely agree, we won't have real location to compare against. How is the accuracy value calculated? I believe we could use HDOP as well to assess accuracy if this value is less reliable.